src/server/master_server/metadata_manager.cc

```diff
@@ -30,53 +54,127 @@ google::protobuf::util::Status MetadataManager::CreateFileMetadata(
 
   absl::WriterMutexLock path_writer_lock_guard(path_lock_or.ValueOrDie());
   
-  // Step 3. Instantiate a FileMetadata object. Use the emplace function 
-  // to detect if the creation took place (as there can be concurrent
-  // creation in rare cases, and only one succeeds). 
-  auto new_file_metadata(std::make_shared<FileMetadata>());
-  // Initialize filename
-  new_file_metadata->set_filename(filename);
-  
-  auto new_file_metadata_and_if_took_place(
-           file_metadata_.emplace(filename, new_file_metadata));
-  auto has_create_taken_place(new_file_metadata_and_if_took_place.second);
+  // Step 3. Instantiate a FileMetadata object.
+  auto map_id(submap_id(filename));
+  absl::Mutex* file_lock(file_metadata_lock_[map_id]);
+  // Acquire a writer lock as we are creating a file metadata in the 
+  // submap underneath
+  absl::WriterMutexLock file_write_lock_guard(file_lock);
 
-  if (!has_create_taken_place) {
+  if (file_metadata_[map_id].contains(filename)) {
     return google::protobuf::util::Status(
         google::protobuf::util::error::ALREADY_EXISTS,
         "File metadata already exists for " + filename);
   }
 
+  auto new_file_metadata(std::make_shared<FileMetadata>());
+  // Initialize filename
+  new_file_metadata->set_filename(filename);
+  // Create the new file metadata in the submap
+  file_metadata_[map_id][filename] = new_file_metadata;  
+
   return google::protobuf::util::Status::OK;
 }
 
 bool MetadataManager::ExistFileMetadata(const std::string& filename) const {
-  return file_metadata_.contains(filename);
+  auto map_id(submap_id(filename));
+  absl::Mutex* file_lock(file_metadata_lock_[map_id]);
```

Comment:
A new class for ThreadSafeFlatHashMap, will encapsulate a lot of this and might make the map consumer code simpler. Will just be a one line map.contains()

Reply:
Yes to some extend. The Exit... method would be feasible to be changed to a one liner such as "map.contains()". But consider the case for Get..., and what you want is return an error if it does not exist or the item itself in an atomic fashion. You cannot do this without exposing the lock, because writing something like:

```
if (!map.contains(...)) {
  return Error;
}

return map.at(...)
```

is buggy because of what can happen in between those lines. That says the necessity of exposing the locks here because what you really want is 

```
lock(sub_map);

if (!map_.contains()) {
  return ERROR;
}

return map_.at(...)
```

Then lets think about the class interface for our use case. We'd have to expose the submap locks, which is ok, but then what is the assumption we make for methods such as "at" or "insert"? We need to keep these operations in the scope of the lock, so the assumption is that caller needs to acquire the submap locks. This can work, but can cause confusion too because user needs to be careful about locking things. 

It is a bit messy, but this is the best thing I can think of at the moment.

Reply:
I see your point. But once we start exposing locks to the caller that kind of defeats the purpose of having the class. The great thing about having our own class is we can define our own interfaces. For your example, we can have our own method: bool TryGetValue(K key, V** pointer_to_pointer_to_val) {
lock(sub_map);

if (!sub_map_.contains()) {
  return false;
}

*pointer_to_pointer_to_val = &sub_map_.at(...); //assuming at() returns value
return true;
}

caller:
if (!map.TryGetValue(...)) {
  handle error;
}
Note: l'm a little bit sleepy, code can be buggy...lol

Something like this, mayn't be the exact code, but just saying we can define our own interfaces for own thread safe map. Thoughts?

Reply:
Maybe we can still do something here to generalize this facility to a class so there is a good code reuse instead of adding ad-hoc treatment here and there. Here is what is my mind, instead of using the intrinsic thread-safe hashmap, let's build our hashmap on top of their default hashmap (so no internal locks), we can define a hashmap that derives from the default hashmap, but keep locks by ourselves. 

```
template<class Key, class Value>
class parallel_hash_map : phmap::parallel_flat_hash_map<Key, Value> {
  public:
  // Fetch the mutex for the submap that corresponds to a key
  absl::Mutex* FetchLock(Key key);
  
  // The remaining methods are inherited from parallel_flat_hash_map  
  private:
     std::vector<shared_ptr<absl::Mutex>> locks_;
}
```

For each key, the submap index can be computed by 
```
   size_t hashval = hash.hash(key);                   // compute its hash
   size_t idx  = hash.subidx(hashval);             // compute the submap index for this hash
```

and the total submap number is hash.subcnt(). 

So put this together, at construction time we should initialize the subcnt() number of locks, and FetchLock function reads as

absl::Mutex* parallel_hash_map::FetchLock(const Key& key) {
   size_t hashhval = hash(key);
   size_t idx = subidx(hashval);
   return lock_[idx].get();
}

Then we can reuse the map, but one needs to always explicitly fetch the lock and perform the following action. 

Does this look like a reasonable revision?

Reply:
okay, it's fine. We can start with this. Thanks.

Reply:
Hi Michael, thanks for your comments. I think we are gradually converging on something. In the reply above, I had suggested extending upon the default parallel hashmap, and I think you brought up a good point that we can have methods such as TryGetValue, TryInsert etc. to take care of the issue I brought up. So we only need a layer of managing the locks ourself. Let me see if I can get this revision going.

Reply:
Sounds great.